src/Icicle/Storage/Dictionary/Toml.hs

```diff
@@ -136,36 +104,36 @@ loadDictionary' parentFuncs parentConf parentConcrete fp
       remakeVirtuals (DictionaryEntry' a (VirtualDefinition' (Virtual' v))) vds = (a, v) : vds
       remakeVirtuals _ vds = vds
 
-sourceCheck :: Monad m
-            => SC.Features () SP.Variable
-            -> SQ.QueryTop Parsec.SourcePos SP.Variable
-            -> EitherT DictionaryImportError m (SQ.QueryTop (ST.Annot Parsec.SourcePos SP.Variable) SP.Variable, ST.Type SP.Variable)
-sourceCheck d q =
-  hoistEither
-       $ (A.left DictionaryErrorCheck)
-       $ snd
-       $ flip Fresh.runFresh (freshNamer "t")
-       $ runEitherT
-       $ SC.checkQT d q
-
-sourceInline :: Dictionary
-             -> SQ.QueryTop (ST.Annot Parsec.SourcePos SP.Variable) SP.Variable
-             -> SQ.QueryTop Parsec.SourcePos SP.Variable
-sourceInline d q
- = SQ.reannotQT ST.annAnnot
- $ inline q
+parseImports conf rp
+ = go `traverse` imports conf
  where
-  funs      = M.map snd
-            $ dictionaryFunctions d
-  inline q' = snd
-            $ Fresh.runFresh
-                (STI.inlineQT funs q')
-                (freshNamer "inline")
-
+  go fp
+   = do let fp'' = T.unpack fp
+        importsText
+          <- EitherT
+           $ A.left DictionaryErrorIO
+          <$> E.try (T.readFile (rp </> fp''))
+        hoistEither
+           $ A.left DictionaryErrorParsecFunc
+           $ P.sourceParseF fp'' importsText
+
+loadImports parentFuncs parsedImports
+ = hoistEither . mapLeft DictionaryErrorCheck
+ $ foldlM go parentFuncs parsedImports
+ where
+  go env f
+   = P.sourceDesugarF f >>= P.sourceCheckF env
 
--- TODO, move this somewhere sensible.
-freshNamer :: T.Text -> Fresh.NameState SP.Variable
-freshNamer prefix = Fresh.counterPrefixNameState (SP.Variable . T.pack . show) (SP.Variable prefix)
+checkDefs d defs
+ = go `traverse` defs
+ where
+  go (a, q)
+   = do  (checked, _)  <- check' d q
+         let inlined    = P.sourceInline d checked
+         (checked', _) <- check' d inlined
+         pure $ DictionaryEntry a (VirtualDefinition (Virtual checked'))
```

Comment:
Looks like the functions are desugared, but do we need to do the Virtuals as well?